spanner/client_test.go

```diff
@@ -392,6 +394,180 @@ func TestClient_MultiEndpoint(t *testing.T) {
+							t.Errorf("TestClient_MultiplexedSession expected multiplexed session to be used, got: %v", req.Session)
+						}
+
+					}
+				}
+			},
+		},
+		{
+			name: "Given if multiplexed session is enabled, When executing multi use R/O transactions, should use multiplexed session",
+			test: func(client *Client) error {
+				ctx := context.Background()
+				// Test the multi use read-only transaction
+				roTxn := client.ReadOnlyTransaction()
+				defer roTxn.Close()
+				iter := roTxn.Query(ctx, NewStatement(SelectSingerIDAlbumIDAlbumTitleFromAlbums))
+				if err := iter.Do(func(row *Row) error {
+					return nil
+				}); err != nil {
+					return err
+				}
+				iter = roTxn.Read(ctx, "Albums", KeySets(Key{"foo"}), []string{"SingerId", "AlbumId", "AlbumTitle"})
+				return iter.Do(func(row *Row) error {
+					return nil
+				})
+			},
+			validate: func(server InMemSpannerServer) {
+				// Validate the multiplexed session is used
+				expected := map[string]interface{}{
+					"SessionsCount": uint(1),
+				}
+				if !isMultiplexEnabled {
+					expected["SessionsCount"] = uint(25) // BatchCreateSession request from regular session pool
+				}
+				if !testEqual(expected["SessionsCount"], server.TotalSessionsCreated()) {
+					t.Errorf("TestClient_MultiplexedSession expected session creation with multiplexed=%s should be=%v, got: %v", strconv.FormatBool(isMultiplexEnabled), expected["SessionsCount"], server.TotalSessionsCreated())
+				}
+				reqs := drainRequestsFromServer(server)
+				for _, s := range reqs {
+					switch s.(type) {
+					case *sppb.ReadRequest:
+						req, _ := s.(*sppb.ReadRequest)
+						// Validate the session is multiplexed
+						if !testEqual(isMultiplexEnabled, strings.Contains(req.Session, "multiplexed")) {
+							t.Errorf("TestClient_MultiplexedSession expected multiplexed session to be used, got: %v", req.Session)
+						}
+					case *sppb.ExecuteSqlRequest:
+						req, _ := s.(*sppb.ExecuteSqlRequest)
+						// Validate the session is multiplexed
+						if !testEqual(isMultiplexEnabled, strings.Contains(req.Session, "multiplexed")) {
+							t.Errorf("TestClient_MultiplexedSession expected multiplexed session to be used, got: %v", req.Session)
+						}
+					}
+
+				}
+			},
+		},
+		{
+			name: "Given if multiplexed session is enabled, When executing R/W transactions, should always use regular session",
+			test: func(client *Client) error {
+				ctx := context.Background()
+				// Test the read-write transaction
+				_, err := client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *ReadWriteTransaction) error {
+					iter := txn.Read(ctx, "Albums", KeySets(Key{"foo"}), []string{"SingerId", "AlbumId", "AlbumTitle"})
+					return iter.Do(func(r *Row) error {
+						return nil
+					})
+				})
+				return err
+			},
+			validate: func(server InMemSpannerServer) {
+				// Validate the regular session is used
+				if !testEqual(uint(25), server.TotalSessionsCreated()) {
```

Comment:
>Note that also here the number of sessions is not deterministic, as we have 4 BatchCreateSessions requests running in parallel.

No we will have 1 parallel request to do BatchCreateSessions, and it will only create that many sessions which are needed(because MinOpened=0 here)